src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs

```diff
@@ -129,31 +110,36 @@ void IDisposable.Dispose()
         /// <param name="inMemoryPdbSize">in memory PDB size</param>
         /// <returns>reader</returns>
         /// <remarks>
-        /// Assumes that neither PE image nor PDB loaded into memory can be unloaded or moved around.
+        /// Accounts for unloadable and dynamic types by keying the cache on the managed Assembly. The
+        /// underlying ConditionalWeakTable doesn't keep the assembly alive, so cached types will be
+        /// correctly invalidated when the Assembly is unloaded by the GC.
         /// </remarks>
-        private unsafe MetadataReader TryGetReader(string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize, IntPtr inMemoryPdbAddress, int inMemoryPdbSize)
+        private unsafe MetadataReader TryGetReader(Assembly assembly, string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize, IntPtr inMemoryPdbAddress, int inMemoryPdbSize)
         {
             if ((loadedPeAddress == IntPtr.Zero || assemblyPath == null) && inMemoryPdbAddress == IntPtr.Zero)
             {
                 // Dynamic or in-memory module without symbols (they would be in-memory if they were available).
                 return null;
             }
 
-            IntPtr cacheKey = (inMemoryPdbAddress != IntPtr.Zero) ? inMemoryPdbAddress : loadedPeAddress;
-
             MetadataReaderProvider provider;
-            while (!_metadataCache.TryGetValue(cacheKey, out provider))
+            while (!_metadataCache.TryGetValue(assembly, out provider))
             {
                 provider = (inMemoryPdbAddress != IntPtr.Zero) ?
                             TryOpenReaderForInMemoryPdb(inMemoryPdbAddress, inMemoryPdbSize) :
                             TryOpenReaderFromAssemblyFile(assemblyPath, loadedPeAddress, loadedPeSize);
 
-                 // If the add loses the race with another thread, then the dispose the provider just 
-                 // created and return the provider already in the cache.
-                 if (_metadataCache.TryAdd(cacheKey, provider))
-                     break;
-
-                 provider?.Dispose();
+                // If the add loses the race with another thread, then the dispose the provider just 
+                // created and return the provider already in the cache.
+                try
```

Comment:
`ConditionalWeakTable` doesn't have a `TryAdd`, so I simply wrapped `Add` in a try/catch.  Would it be worth adding this functionality to `ConditionalWeakTable`?

Reply:
No longer necessary in light of using `ConditionalWeakTable`'s `GetValue(key, (key) => value)` override which atomically creates values.